### ISSUE TITLE
Support cred plugins in the "twistd ftp" command line

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4752]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4752|
|Type|enhancement|
|Created|2010-12-03 03:28:59Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-AuthOptionMixin-4752-2|

The FTP server would be 31% more awesome if it supported pluggable authentication backends, instead of only supporting a password file.  It should use `twisted.cred.strcred.AuthOptionMixin` instead of its own `--password-file` thing.

It would be nice if `--userAnonymous` could also be eliminated, with the determination about whether a user is anonymous delegated to the credentials checker.  However, it appears as though the server wants to return "331.2" instead of "331.1" when an anonymous login is being attempted, which requires the server to know if it is going to be an anonymous login before the "password" is sent (hence before the login is attempted).

<details><summary>Searchable metadata</summary>

```
trac-id__4752 4752
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_ftp_AuthOptionMixin_4752_2 branches/ftp-AuthOptionMixin-4752-2
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1291346939000000 1291346939000000
changetime__1319963220000000 1319963220000000
version__None None
owner__cyli cyli

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Perhaps we can get rid of `--userAnonymous`.

After FTP receives the username, it can log in to the portal with `UsernamePassword(username, "")`.  If this succeeds, the server can proceed to prompt for the "password" (ie, email address) and log it or whatever.  If it fails, the server can prompt for the password and re-try login with `UsernamePassword(username, password)`.

This would let an `IUsernamePassword` credentials checker decide which username corresponds to the anonymous user and return `ANONYMOUS` when it sees a login attempt for that username.  If the realm also allows anonymous users, this would succeed with a usable avatar.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli commented|
|-|-|

(In [32844]) Branching to 'ftp-AuthOptionMixin-4752'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli commented|
|-|-|

(In [32853]) Add cred plugin support to twistd ftp and deprecate --password-file option

refs [#4752](https://github.com/twisted/twisted/issues/4752)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ftp-AuthOptionMixin-4752)

Moved the --userAnonymous stuff to [#5313](https://github.com/twisted/twisted/issues/5313)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

(In [32976]) Forgot to add checkers to portal  refs [#4752](https://github.com/twisted/twisted/issues/4752)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_8"></a>|@cyli commented|
|-|-|

[build results again](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ftp-AuthOptionMixin-4752)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to @cyli|
|-|-|

Thanks.  This is an area of work we really need to devote some attention to, so
thanks for getting the ball rolling here :-).

    1. `.misc` files should generally be empty; I think that this deserves a
    `4752.feature` and a `4752.removal` at the same time, since this is adding a
    feature and removing something at the same time.  (I seem to remember that
    there's a precedent for this, but I can't remember where I saw it.  Maybe
    exarkun can recall.)
    2. For some reason `test_ftp_options.py` and the `.misc` file don't have a
    newline at the end, which makes the diffs look a little weird.
    3. There's some trailing whitespace in `test_ftp_items.py`.
    4. You need to resolve some merge conflicts, so please merge forward and
    re-run the tests.
    5. There are a bunch of pyflakes errors, but they're all imported-but-unused:
```
twisted/tap/ftp.py:11: 'error' imported but unused
twisted/tap/ftp.py:11: 'credentials' imported but unused
twisted/tap/ftp.py:16: 'os' imported but unused
twisted/test/test_ftp_options.py:10: 'UsageError' imported but unused
twisted/test/test_ftp_options.py:11: 'makeService' imported but unused
twisted/test/test_ftp_options.py:12: 'deprecate' imported but unused
```
    6. `opt_` methods are slightly special, since their docstrings are displayed
    as part of `--help`.  So you shouldn't put `@since` markers (or any other
    epytext, really) in them.  (Whenever you edit a tap plugin, you should
    probably have a look at their help interactively, e.g. `twistd ftp --help`.)
    7. `testPasswordfileDeprecation` should be `test_passwordfileDeprecation`;
    `test_` is a method prefix.
    8. We try to avoid the word "correct" in test docstrings, since it doesn't
    actually say what is being verified.  Instead, describe the property
    involved.  Also, we don't describe what the test is testing, but just the
    property of the system is.  So, for example, instead of saying "Test that
    the --passwordfile option will emit a correct warning", instead, "The
    --passwordfile option emits a warning stating that said option is
    deprecated."
    9. The docstring for `FTPOptionsTestCase` looks copy/pasted: it says it's
    testing `twistd mail` but I think it's pretty obvious it's testing `twistd
    ftp`.  Also, the `C{}` epytext markup ("code") would be more appropriate
    than `I{}` ("italic") for annotating a shell command.
    10. This branch does involve new functionality for which there are no new
    tests: there really needs to be at least one test that invokes the new
    `--auth` option and makes sure it actually does something; after all, the
    old tests pass just fine if the new `credCheckers` field is totally ignored
    and not passed to the `Portal` at all.

Thanks again; aside from item 10, none of this is really functional, so it's
_almost_ ready to land; I suspect the next review will be very brief.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_10"></a>|@cyli commented|
|-|-|

(In [33006]) Branching to 'ftp-AuthOptionMixin-4752-2'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_11"></a>|@cyli commented|
|-|-|

(In [33007]) Pyflakes and trailing whitespace and other changes refs [#4752](https://github.com/twisted/twisted/issues/4752)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_12"></a>|@cyli commented|
|-|-|

(In [33008]) Add auth mixin tests refs [#4752](https://github.com/twisted/twisted/issues/4752)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_13"></a>|@cyli removed owner|
|-|-|

[new build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ftp-AuthOptionMixin-4752-2)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

Reviewing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph set owner to @cyli<br>@glyph set status to `new`|
|-|-|

Looks pretty good.  A few things you may want to tweak before merging:

    1. The 'longdesc' attribute talks about creating a `.tap` file, which we
    haven't done for many years.  That should probably just be restored to be
    `''`.  Looks like this may have been copied from `twisted.mail.tap`'s
    example?
    2. It looks like you didn't quite respond to review point 8.  "Test that"
    and "correct" are in the same genre of things not to put in test docstrings.
    Just describe the property being verified in the test docstrings, not the
    method of its verification.  Also, be as specific as possible about the
    properties of correct/incorrect, valid/invalid inputs and outputs.  For
    example, instead of:
```
        Tests that the --auth command generates a checker does not authorize
        invalid logins
```
    what about:
```
        The C{--auth} command-line option add a checker to the list of checkers,
        which returns a L{Deferred} that fails with L{UnauthorizedLogin} when
        presented with credentials that are unknown to that checker.
```
    This isn't the best possible docstring, but it says exactly what the
    expected behavior is, rather than having the user guess what "authorize" and
    "valid credentials" means.
    3. `test_passwordfileDeprecation` refers to the `--passwordfile` option in
    its docstring, but the actual option is `--password-file`.
    4. I notice that you're using the `file:` strcred plugin and writing a
    temporary file in these tests, but you could use `memory:` and skip that
    step.
    5. It's a good idea to avoid returning `Deferred`s from test methods unless
    you need to process real events in your tests.  In the case of
    `test_ftp_options`, you know
    exactly what is supposed to happen: the Deferred returned by the
    `requestAvatarId` method for the `file:` plugin should always return
    something synchronously.  If the `Deferred` returned by `requestAvatarId`
    ''doesn't'' have a result immediately, you want the test to fail immediately
    and report that problem, not hang forever and require the user (or the
    buildbot) to time out the test or kill it with `^C`.  So, for example,
    instead of this:
```
#!py
return checker.requestAvatarId(correct).addCallback(
    lambda username: self.assertEqual(username, correct.username)
)
```
    do something like this:
```
#!py
usernames = []
checker.requestAvatarId(correct).addCallback(usernames.append)
self.assertEqual(usernames, [correct.username])
```

Address these points to your satisfaction and then land.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_16"></a>|@cyli commented|
|-|-|

(In [33009]) Change docstrings to match review comments, and clear out longdesc
refs [#4752](https://github.com/twisted/twisted/issues/4752)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_17"></a>|@cyli set status to `closed`|
|-|-|

(In [33010]) Merge ftp-AuthOptionMixin-4752-2 : add cred plugin to ftp tap

Author: cyli
Reviewer: glyph
Fixes: [#4752](https://github.com/twisted/twisted/issues/4752)

Add support for cred plugins in twistd ftp command line
